### PR TITLE
fix(db pull): --url no longer overrides the datasource provider

### LINKED ISSUE #11527: Using `--url` for `db pull` ignored/overwrites `provider=cockroachdb`
With this schema:
```prisma
generator client {
  provider        = "prisma-client-js"
  previewFeatures = ["cockroachdb"]
}

datasource db {
  provider = "cockroachdb"
  url      = "postgresql://jo_l:11...oud:26257/introspection-ci-5134.aragon-notification-service?sslmode=verify-full"
}
```

this is the output of `prisma db pull --url=postgresql://jo....cloud:26257/introspection-ci-5134.aragon-notification-service?sslmode=verify-full`:
```
λ npx prisma db pull --url=postgresql://jo_l:1...oud:26257/introspection-ci-5134.aragon-notification-service?sslmode=verify-full
Environment variables loaded from .env
Prisma schema loaded from prisma\schema.prisma
Datasource "db": CockroachDB database "introspection-ci-5134.aragon-notification-service", schema "public" at "free-tier...ud:26257"

Introspecting …

✔ Introspected 6 models and wrote them into prisma\schema.prisma in 6.80s

Run prisma generate to generate Prisma Client.
```
(which is great!)

but this the resulting schema:

```
generator client {
  provider        = "prisma-client-js"
  previewFeatures = ["cockroachdb"]
}

datasource db {
  provider = "postgresql"
  url      = "postgresql://jo_l:11ekBxMQuUr07kHv@free-tier.gcp-us-central1.cockroachlabs.cloud:26257/introspection-ci-5134.aragon-notification-service?sslmode=verify-full"
}

model eventsources {
  eventsource_id   BigInt          @id
  created_at       DateTime?       @default(now()) @db.Timestamptz(6)
  enabled          Boolean         @default(true)
  contract_address String          @db.VarChar(255)
```

Note how it says `provider = "postgresql"`  there.

I think it would be ok to keep ´provider=cockroachdb` around when already set in the original schema.

(Everything works fine when you just run `db pull` without `--url`)

Related: https://github.com/prisma/prisma/issues/11347

### PR DESCRIPTION
**Pull request by jkomyno**
Closes https://github.com/prisma/prisma/issues/11527.

### CODE REVIEW

**Comment by jkomyno** [comment]
We currently have 45 successful checks. Let's see how it goes using the `cockroachdb` service

**Comment by jkomyno** [comment]
In order to fix the tests in this PR, we need:
- a valid `.github/setup-cockroachdb.sh` script, with a related "Setup CockroachDB" CI step for the "no-docker" workflow
- understand how to make sure the same snapshots that work for Github Actions also work for Buildkite

cc @janpio @Jolg42

**Comment by janpio** [comment]
I think you can skip the Cockroach related tests in no-docker - even the Docker configuration for CockroachDB is still very much in flux and will change in the coming does.

What exactly is the problem with the snapshots?

**Comment by jkomyno** [comment]
The problem with snapshots in Buildkite is that the URL with `cockroachdb` as host is printed, whereas in Github Actions the `localhost` URL is printed, but if I change the "COCKROACH" env var in Buildkite to the same used in Github Actions, I get a connection error. Please see the last 2/3 commits for further details

**Comment by janpio** [comment]
Ok, that difference between the docker-compose files for Buildkite and GH Actions always confuses me as well. I do not know how we deal with that usually, but @Jolg42 probably does and can help out here. (If all fails we can sanitize it away in the snapshot)

File: .buildkite/test/docker-compose.14.yml [hunk 1]
```diff
@@ -31,7 +31,7 @@ services:
       - TEST_MONGO_URI=mongodb://root:prisma@mongo:27018/tests?authSource=admin
       - TEST_MONGO_URI_MIGRATE=mongodb://root:prisma@mongodb_migrate:27017/tests-migrate?authSource=admin
       # CockroachDB
-      - TEST_COCKROACH_URI=postgresql://prisma@cockroachdb:26257/
+      - TEST_COCKROACH_URI=postgresql://prisma@localhost:26257/tests
```
Comment (Jolg42) [inline .buildkite/test/docker-compose.14.yml:34]:
```suggestion
      - TEST_COCKROACH_URI=postgresql://prisma@cockroachdb:26257/tests
```
Host needs to be `cockroachdb` in Buildkite (name of the docker-compose step)

File: .buildkite/publish/docker-compose.yml [hunk 2]
```diff
@@ -31,7 +31,7 @@ services:
       - TEST_MONGO_URI=mongodb://root:prisma@mongo:27018/tests?authSource=admin
       - TEST_MONGO_URI_MIGRATE=mongodb://root:prisma@mongodb_migrate:27017/tests-migrate?authSource=admin
       # CockroachDB
-      - TEST_COCKROACH_URI=postgresql://prisma@cockroachdb:26257/
+      - TEST_COCKROACH_URI=postgresql://prisma@localhost:26257/tests
```
Comment (Jolg42) [inline .buildkite/publish/docker-compose.yml:34]:
```suggestion
      - TEST_COCKROACH_URI=postgresql://prisma@cockroachdb:26257/tests
```
Host needs to be `cockroachdb` in Buildkite (name of the docker-compose step)

File: packages/migrate/src/__tests__/__snapshots__/DbPull.test.ts.snap [hunk 3]
```diff
@@ -66,6 +66,121 @@ model jobs {
 }
 
 
+// introspectionSchemaVersion: NonPrisma,
+`;
+
+exports[`cockroachdb basic introspection --url 2`] = `
+datasource db {
+  provider = "postgresql"
+  url      = "postgresql://prisma@localhost:26257/tests"
```
Comment (Jolg42) [inline packages/migrate/src/__tests__/__snapshots__/DbPull.test.ts.snap:75]: The value here will be `cockroachdb` instead of `localhost` for Buildkite so we can normalize this here 
https://github.com/prisma/prisma/blob/main/packages/sdk/src/utils/jestSnapshotSerializer.js#L73:L75

Comment (jkomyno) [reply packages/migrate/src/__tests__/__snapshots__/DbPull.test.ts.snap:75]: Thanks, this was the sanitization part I was missing yesterday!

Comment (jkomyno) [reply packages/migrate/src/__tests__/__snapshots__/DbPull.test.ts.snap:75]: @Jolg42 still, we need a `.github/setup-cockroachdb.sh` file. I'll try to write one, but I'm not that familiar with cockroachdb

File: packages/migrate/src/__tests__/DbPull.test.ts [hunk 4]
```diff
@@ -491,6 +491,64 @@ describe('postgresql', () => {
...
+    await expect(result).resolves.toMatchInlineSnapshot(``)
+    expect(ctx.mocked['console.log'].mock.calls.join('\n')).toMatchSnapshot()
+    expect(ctx.mocked['console.info'].mock.calls.join('\n')).toMatchInlineSnapshot(``)
+    expect(ctx.mocked['console.error'].mock.calls.join('\n')).toMatchInlineSnapshot(``)
+  })
+
+  // TODO: Windows: fails with
+  // Error: P1012 Introspection failed as your current Prisma schema file is invalid·
+  //     Please fix your current schema manually, use prisma validate to confirm it is valid and then run this command again.
+  //     Or run this command with the --force flag to ignore your current schema and overwrite it. All local modifications will be lost.
```
[the remark this answers is not in the document]
Comment (jkomyno) [reply packages/migrate/src/__tests__/DbPull.test.ts:540]: Both. Copy-paste of another test case (just plugging cockroachdb in it) which happens to fail with that error on Windows

File: packages/migrate/src/commands/DbPull.ts [hunk 5]
```diff
@@ -144,22 +146,33 @@ Set composite types introspection depth to 2 levels
...
-    } else if (schemaPath) {
-      schema = fs.readFileSync(schemaPath, 'utf-8')
-    } else {
-      throw new Error('Could not find a `schema.prisma` file')
-    }
+          if (input.url) {
+            const config = await getConfig({
+              datamodel: rawSchema,
+              ignoreEnvVarErrors: true,
+            })
```
Comment (Jolg42) [inline packages/migrate/src/commands/DbPull.ts:159]: Note: one more place where we use `getConfig()` @janpio

File: packages/migrate/src/commands/DbPull.ts [hunk 6]
```diff
@@ -231,6 +244,9 @@ Then you can run ${chalk.green(getCommandWithExecutor('prisma db pull'))} again.
       } else if (e.code === 'P1012') {
         // Schema Parsing Error
         console.info() // empty line
+
+        // TODO: this error is misleading, as it gets thrown even when the schema is valid, put the protocol of the given
+        // '--url' argument is different than the one written in the schema.prisma file.
```
Comment (Jolg42) [inline packages/migrate/src/commands/DbPull.ts:249]: Do you have a reproduction? Creating a small issue would be good to track this 📝

File: packages/migrate/src/commands/DbPull.ts [hunk 7]
```diff
@@ -144,22 +146,33 @@ Set composite types introspection depth to 2 levels
...
-    } else {
-      throw new Error('Could not find a `schema.prisma` file')
-    }
+          if (input.url) {
+            const config = await getConfig({
+              datamodel: rawSchema,
+              ignoreEnvVarErrors: true,
+            })
+            const provider = config.datasources[0]?.provider
+            const schema = this.urlToDatasource(input.url, provider) + removeDatasource(rawSchema)
```
Comment (Jolg42) [inline packages/migrate/src/commands/DbPull.ts:161]:
```suggestion
            const schema = `${this.urlToDatasource(input.url, provider)}${removeDatasource(rawSchema)}`
```
Usually, it's a safer default to go for template string concatenation
It should do the same in the end.

File: packages/migrate/src/commands/DbPull.ts [hunk 8]
```diff
@@ -144,22 +146,33 @@ Set composite types introspection depth to 2 levels
       throw new NoSchemaFoundError()
     }
 
-    let schema: string | null = null
+    const schema = await match({ url, schemaPath })
+      .when(
```
Comment (Jolg42) [inline packages/migrate/src/commands/DbPull.ts:150]: Could use some small comments about the different cases maybe, so it's easier to read without going to the details?

File: packages/migrate/src/commands/DbPull.ts (hunk 7 above)
Comment (jkomyno) [reply packages/migrate/src/commands/DbPull.ts]: Right, check the new commit!

File: packages/migrate/src/commands/DbPull.ts (hunk 8 above)
Comment (jkomyno) [reply packages/migrate/src/commands/DbPull.ts:162]: Right, check the new commit!

**Comment by Jolg42** [comment]
a test fails with 
```

  ● cockroachdb › basic introspection --url

    expect(received).resolves.toMatchInlineSnapshot()

    Received promise rejected instead of resolved
    Rejected to value: [Error: [introspection-engine/connectors/sql-introspection-connector/src/version_checker.rs:78:97] called `Result::unwrap()` on an `Err` value: Error("unknown variant `String`, expected one of `SmallInt`, `Integer`, `BigInt`, `Decimal`, `Money`, `Inet`, `Oid`, `Citext`, `Real`, `DoublePrecision`, `VarChar`, `Char`, `Text`, `ByteA`, `Timestamp`, `Timestamptz`, `Date`, `Time`, `Timetz`, `Boolean`, `Bit`, `VarBit`, `Uuid`, `Xml`, `Json`, `JsonB`", line: 0, column: 0)]
```

Might be since the merge of main in this PR

**Review by Jolg42 — APPROVED** [review]
Nice PR! Last step would be to fix the test error and then merging time :shipit:

File: packages/migrate/src/commands/DbPull.ts (hunk 6 above)
Comment (jkomyno) [reply packages/migrate/src/commands/DbPull.ts]: Here is how to reproduce:

- Initialize a new schema with e.g. the `mongodb` provider: `npx prisma init --datasource-provider mongodb`
- Ensure the schema is valid: `npx prisma validate`
- Run introspection with a URL having a different datasource protocol: `npx prisma db pull --url postgres://prisma:prisma@localhost:5432/tests`

You should see an error similar to the following:

```text
Prisma schema loaded from prisma/schema.prisma
Environment variables loaded from .env
Datasource "db"

✖ Introspecting

Error: P1012 Introspection failed as your current Prisma schema file is invalid

Please fix your current schema manually, use prisma validate to confirm it is valid and then run this command again.
Or run this command with the --force flag to ignore your current schema and overwrite it. All local modifications will be lost.
```

**Comment by Jolg42** [comment]
Tests should pass after https://github.com/prisma/prisma-engines/pull/2872 is merged

File: packages/migrate/src/__tests__/__snapshots__/DbPull.test.ts.snap [hunk 9]
```diff
@@ -146,6 +181,45 @@ enum Role {
...
+  coinflips Boolean[]
+  User      User?     @relation(fields: [authorId], references: [id])
+}
+
+model User {
+  id    String  @id
+  email String  @unique(map: "User.email")
+  name  String?
+  Post  Post[]
+}
```
Comment (janpio) [inline packages/migrate/src/__tests__/__snapshots__/DbPull.test.ts.snap:215]: We should probably make some of the `String` fields different native types, that are different between CockroachDB and PostgreSQL provider.

Comment (jkomyno) [reply packages/migrate/src/__tests__/__snapshots__/DbPull.test.ts.snap:215]: So, in this case they should be `VarChars`, right? Since we have a postgresql schema

**Review by Jolg42 — CHANGES_REQUESTED** [review]
We discussed that we might want to go through some things here before merging.

File: packages/migrate/src/commands/DbPull.ts [hunk 10]
```diff
@@ -231,6 +256,9 @@ Then you can run ${chalk.green(getCommandWithExecutor('prisma db pull'))} again.
       } else if (e.code === 'P1012') {
         // Schema Parsing Error
         console.info() // empty line
+
+        // TODO: this error is misleading, as it gets thrown even when the schema is valid but the protocol of the given
+        // '--url' argument is different than the one written in the schema.prisma file.
```
[the remark this answers is not in the document]
Comment (jkomyno) [reply packages/migrate/src/commands/DbPull.ts:261]: Here is how to reproduce:

- Initialize a new schema with e.g. the `mongodb` provider: `npx prisma init --datasource-provider mongodb`
- Ensure the schema is valid: `npx prisma validate`
- Run introspection with a URL having a different datasource protocol: `npx prisma db pull --url postgres://prisma:prisma@localhost:5432/tests`

You should see an error similar to the following:

```text
Prisma schema loaded from prisma/schema.prisma
Environment variables loaded from .env
Datasource "db"

✖ Introspecting

Error: P1012 Introspection failed as your current Prisma schema file is invalid

Please fix your current schema manually, use prisma validate to confirm it is valid and then run this command again.
Or run this command with the --force flag to ignore your current schema and overwrite it. All local modifications will be lost.
```

[the remark this answers is not in the document]
Comment (janpio) [reply packages/migrate/src/commands/DbPull.ts:261]: Ok, then my analysis would be that the combination of `provider=mongodb` and the `--url` value lead to validation failing before Introspeciton is run. (Similar how setting an env var with a mismatched `protocol` would lead to a similar problem)

File: packages/migrate/src/__tests__/DbPull.test.ts [hunk 11]
```diff
@@ -58,8 +59,7 @@ describe('common/sqlite', () => {
     expect(ctx.mocked['process.stderr.write'].mock.calls.join('\n')).toMatchInlineSnapshot(``)
   })
 
-  test('basic introspection with invalid --url', async () => {
-    ctx.fixture('introspection/sqlite')
+  test('basic introspection with invalid --url if schema is unspecified', async () => {
```
Comment (janpio) [inline packages/migrate/src/__tests__/DbPull.test.ts:62]: why this change and removal of fixtures on that existing test?

File: .buildkite/publish/docker-compose.yml [hunk 12]
```diff
@@ -31,7 +31,7 @@ services:
       - TEST_MONGO_URI=mongodb://root:prisma@mongo:27018/tests?authSource=admin
       - TEST_MONGO_URI_MIGRATE=mongodb://root:prisma@mongodb_migrate:27017/tests-migrate?authSource=admin
       # CockroachDB
-      - TEST_COCKROACH_URI=postgresql://prisma@cockroachdb:26257/
+      - TEST_COCKROACH_URI=postgresql://prisma@cockroachdb:26257/tests
```
Comment (janpio) [inline .buildkite/publish/docker-compose.yml:34]: Is this env var already used anywhere in code? 
(If thought there were not CockroachDB tests - why do we have the env var then?)

File: packages/migrate/src/__tests__/DbPull.test.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (janpio) [reply packages/migrate/src/__tests__/DbPull.test.ts:550]: That indeed sounds like something we need to figure out still.

File: .buildkite/publish/docker-compose.yml (hunk 12 above)
Comment (jkomyno) [reply .buildkite/publish/docker-compose.yml:34]: This PR is the first that introduces cockroachdb tests

File: packages/migrate/src/__tests__/DbPull.test.ts (hunk 11 above)
Comment (jkomyno) [reply packages/migrate/src/__tests__/DbPull.test.ts:62]: Because the schema is automatically inferred if `prisma/prisma.schema` is available in the current location. We wanted to test the case in which the `--url` flag is executed without any knowledge of a prisma schema.

File: packages/migrate/src/__tests__/DbPull.test.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (jkomyno) [reply packages/migrate/src/__tests__/DbPull.test.ts:550]: Agree, but in another PR imho. The same "Windows fails with error P1012" is excluded from the CI for other databases in other test cases

[the remark this answers is not in the document]
Comment (janpio) [reply packages/migrate/src/__tests__/DbPull.test.ts:550]: Ok then please create an issue and optimally link it in the TODO

File: packages/migrate/src/__tests__/DbPull.test.ts (hunk 11 above)
Comment (janpio) [reply packages/migrate/src/__tests__/DbPull.test.ts:62]: And the previously existing test is useless and is not needed any more?

File: packages/migrate/src/commands/DbPull.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (janpio) [reply packages/migrate/src/commands/DbPull.ts:261]: Can you open an issue for this TODO as well pls?

[the remark this answers is not in the document]
Comment (jkomyno) [reply packages/migrate/src/commands/DbPull.ts:261]: https://github.com/prisma/prisma/issues/13076

File: packages/migrate/src/__tests__/DbPull.test.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (jkomyno) [reply packages/migrate/src/__tests__/DbPull.test.ts:550]: Opened https://github.com/prisma/prisma/issues/13077.

File: packages/migrate/src/__tests__/__snapshots__/DbPull.test.ts.snap (hunk 9 above)
Comment (jkomyno) [reply packages/migrate/src/__tests__/__snapshots__/DbPull.test.ts.snap:215]: On a closer inspection, the way these models are generated is not dependent on the code shipped with this PR, as both the postgres schema and the cockroach one have no model defined. We are merely introspecting existing tables in the postgres and cockroach images running on Docker. I have just merged `main` in `integration/fix-db-pull-url-overwrites-provider-cockroachdb` to trigger the CI, let's see if anything changes.

Comment (jkomyno) [reply packages/migrate/src/__tests__/__snapshots__/DbPull.test.ts.snap:215]: The CI succeeds nonetheless, I suggest we open a new issue to verify why the introspection results in "general types" rather than "native types" (if that's unexpected at all), and merge this PR in the meantime.
Do you agree? @janpio @Jolg42

File: packages/migrate/src/__tests__/DbPull.test.ts [hunk 13]
```diff
@@ -491,6 +491,74 @@ describe('postgresql', () => {
...
+  test('basic introspection (no schema) --url', async () => {
+    const introspect = new DbPull()
+    const result = introspect.parse(['--print', '--url', setupParams.connectionString])
+    await expect(result).resolves.toMatchInlineSnapshot(``)
+    expect(ctx.mocked['console.log'].mock.calls.join('\n')).toMatchSnapshot()
+    expect(ctx.mocked['console.info'].mock.calls.join('\n')).toMatchInlineSnapshot(``)
+    expect(ctx.mocked['console.error'].mock.calls.join('\n')).toMatchInlineSnapshot(``)
+  })
+
+  // TODO: Windows: fails with
```
Comment (janpio) [inline packages/migrate/src/__tests__/DbPull.test.ts:547]:
```suggestion
  // TODO: (https://github.com/prisma/prisma/issues/13077) Windows: fails with
```

File: packages/migrate/src/__tests__/__snapshots__/DbPull.test.ts.snap [hunk 14]
```diff
@@ -66,6 +66,160 @@ model jobs {
 }
 
 
+// introspectionSchemaVersion: NonPrisma,
+`;
+
+exports[`cockroachdb basic introspection (no schema) --url 2`] = `
+datasource db {
+  provider = "postgresql"
```
Comment (janpio) [inline packages/migrate/src/__tests__/__snapshots__/DbPull.test.ts.snap:74]: Is this expected to be PostgreSQL?

Comment (janpio) [reply packages/migrate/src/__tests__/__snapshots__/DbPull.test.ts.snap:74]: Yes, as we only use the provider protocol or the existing provider in the schema file.

**Review by Jolg42 — APPROVED** [review]
💯 Here we gooo
